File: leetcode/addTwoNumbers/addTwoNumbers.cpp
// Definition for singly-linked list.
#include <iostream>
#include <vector>
#include <math.h>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(long x) : val(x), next(NULL) {}
};
 
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode dummy(0);
        ListNode* p = &dummy;
        int cn = 0;
        int val = 0;
        while(l1 || l2){
            val = cn + (l1 ? l1->val : 0) + (l2 ? l2->val : 0);
            cn = val / 10;
            val = val % 10;
            p->next = new ListNode(val);
            p = p->next;
            if(l1)
                l1 = l1->next;
            if(l2)
                l2 = l2->next;
        }

        if(cn){
            p->next = new ListNode(cn);
            p = p->next;
        }

        return dummy.next;
    }
};

int main(){
    vector<int> l1_vals{9};
    vector<int> l2_vals{1,9,9,9,9,9,9,9,9,9};
    ListNode* l1 = new ListNode(9);
    int n = l2_vals.size();
    // cout << n << endl;
    ListNode* l2 = new ListNode(1);
    ListNode* flag=l2;
    ListNode* temp(nullptr);

    for(int i=1; i< n; i++){
        temp = new ListNode(l2_vals[i]);
        flag->next = temp;
        flag = temp;
    }
    flag = l2;
    int val;
    // do{
    //     val = flag->val;
    //     cout << val << endl;
    // } while(flag = flag->next);
    
    Solution solver;
    ListNode* re = solver.addTwoNumbers(l1, l2);

    flag = re;
    do{
        val = flag->val;
        cout << val << endl;
    } while(flag = flag->next);
    return 0;

}

File: leetcode/addTwoNumbers/addTwoNumbers.py
# You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order and 
# each of their nodes contain a single digit. Add the two numbers and return it as a linked list.
# You may assume the two numbers do not contain any leading zero, except the number 0 itself.
# Example

# Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
# Output: 7 -> 0 -> 8
# Explanation: 342 + 465 = 807.


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        order = 1
        val_a = l1.val
        node_next = l1.next
        while(node_next):
            val_a += node_next.val * (10**order) 
            node_next = node_next.next
            order += 1
        
        order = 1
        val_b = l2.val
        node_next = l2.next
        while(node_next):
            val_b += node_next.val * (10**order) 
            node_next = node_next.next
            order += 1
        
        total = val_a + val_b
        vals = []
        mod = total % 10
        vals.append(mod)
        total = total // 10
        while(total):
            mod = total % 10
            vals.append(mod)
            total = total // 10

        list_node = []
        for val in vals:
            list_node.append(ListNode(val))
        for i in range(len(list_node) -1):
            list_node[i].next = list_node[i+1]
        return list_node[0]    


        

